Sim tool refactor

Maelstromeous opened the pull request:
Objective of this PR is to refactor the sim tool into reusable classes and constants enabling the ability to rapidly create new scenarios. This also has the added benefit of keeping the code very clean and to the point, using defaults to classes where appropiate and making use of dataclasses, in particular `dataclass_json` to handle the encoding of classes into JSON.

Outstanding tasks:

- [ ] Use the rabbitmq exchange rather than the individual queues to ensure full end to end testing of message routing
- [ ] Create DeathEvent generator for allow configurable number of events to be issued and asserted

Commit:
WIP: Refactored the entire testing suite to extrapolote as much as possible into classes

Commit:
Added mapVersion check

Commit:
Now successfully building a metagame message ready for publishing

Commit:
Brought death event back in line

Commit:
Now successfully publishing message to metagame event queue

Commit:
Now successfully publishing message to metagame event queue

Commit:
Fixed metagame event not accepting the timestamp format

Commit:
Now providing correct formatted message to the queue

Commit:
Added ending of instance with delay

Commit:
Added Ps2AlertsApiOps which enables us to make PS2Alerts API requests very easily

Commit:
Removed instance by accident

Commit:
Added logger. Added assertion of ended metagame instance

Commit:
Fixed colours on the formatter

Commit:
Slight refactor to abstract rabbit sending messages to a new class

ryanjsims commented on alert-sim/dataclass/OutfitwarsInstance.py:
```diff
@@ -0,0 +1,32 @@
+    round: int
+    instanceId: str = ''
+    timeEnded: Optional[datetime] = None
+    result: MetagameResult = MetagameResult(33, 33, 33, 0, 0)
```
The default here should be MetagameResult(0, 33, 33, 33, 0) since there won't be any purple territory and 3 of the bases start out neutral for an outfit-wars instance

Maelstromeous replied:
Ah yeah this was a copy pasta doodoo

ryanjsims commented on alert-sim/operations/MetagameEventOps.py:
```diff
@@ -0,0 +1,59 @@
+from pika.adapters.blocking_connection import BlockingChannel
+from time import sleep
+import sys
+sys.path.append("..") # Adds higher directory to python modules path. This is so dumb.
```
I think there's a better way to do this, probably by making everything you want to import here a fully fledged python package and installing it in a virtual env or pipenv

ryanjsims commented on alert-sim/operations/PreflightChecksOps.py:
```diff
@@ -0,0 +1,12 @@
+import sys
+sys.path.append("..") # Adds higher directory to python modules path. This is so dumb.
```
Same as above

ryanjsims commented on alert-sim/operations/Ps2AlertsApiOps.py:
```diff
@@ -0,0 +1,26 @@
+import sys
+sys.path.append("..") # Adds higher directory to python modules path. This is so dumb.
```
And here

ryanjsims commented on alert-sim/operations/RabbitOps.py:
```diff
@@ -0,0 +1,32 @@
+from pika.adapters.blocking_connection import BlockingChannel
+import sys
+sys.path.append("..") # Adds higher directory to python modules path. This is so dumb.
+from typing import Dict
```
Another place that the package change would fix

ryanjsims reviewed (commented):
Added some suggestions on how to fix the path issue as well as a default adjustment suggestion - I'll look into making packages out of everything

Commit:
Added WIP of DeathEvent publishing

Commit:
Removed old comment

ryanjsims commented on alert-sim/dataclass/FacilityControl.py:
```diff
@@ -0,0 +1,22 @@
+    mapControl: Optional[MapControl] = None
+
+    def __post_init__(self):
+        self.timestamp = self.timestamp.isoformat()
```
I would not recommend converting this to a string if the type on the class is datetime

Maelstromeous replied:
@ryanjsims unfortunately it breaks JSON encoding so we have no real choice - JSON can't seem to parse the datetime object so I had to make it a string

ryanjsims commented on alert-sim/operations/MetagameEventOps.py:
```diff
@@ -0,0 +1,58 @@
+        metagameEvent = MetagameEvent(
+            str(instance.censusInstanceId),
+            str(MetagameEventType.INDAR_ENLIGHTENMENT.value),
+            str(MetagameEventState.STARTED.value),
```
Is there a reason its using the STARTED state? Or is this a typo?

Maelstromeous replied:
Ironically it should be the ended state but it didn't seem to make any difference 🧐 but yeah it should be ended

ryanjsims committed:
Move away from pipenv to just use normal virtual environments

ryanjsims committed:
Clean up a bunch of stuff

Fixed the relative import/adding .. to path issue by making alert-sim a package

Added Facility Control event

Made rabbit its own service so you can simply call rabbit.send(event, queuename)
rather than passing around connections

Moved some classes out of outfit-wars-match.py

added __all__ lists to most packages (for from X import * syntax)

ryanjsims committed:
Working on fixes

ryanjsims committed:
Reformat logs

ryanjsims committed:
Updated events to auto set timestamp correctly

ryanjsims committed:
fixed map issues

ryanjsims committed:
Fix issue where match would not end when alpha or omega was captured

ryanjsims committed:
Add team id to DeathEvents

ryanjsims committed:
Outfits on the wrong team may not get credit for captures

Commit:
Wrapped rake matches in a shell

Commit:
Added missing python packages from requirements.txt, corrected fat finger

Commit:
WIP death generator

Commit:
Corrected my character ID, I think it was the DIG outfit one before

ryanjsims committed:
Add census integration to better simulate outfit war

Also adds more switches and dials to adjust how the match plays

ryanjsims committed:
Add usage after runOW.sh refactor

ryanjsims commented:
Some outstanding tasks that need done:

- [ ] Use the rabbitmq exchange rather than the individual queues to ensure full end to end testing of message routing

ryanjsims committed:
Add full outfitwar simulation tool

Commit:
Fixed the tool thinking you've not supplied a service ID when you have

ryanjsims committed:
Correct win condition and update rankings after championship

Commit:
Created runOWMatches to wrap dispatcher
Corrected a int parse parsing an int while you int

ryanjsims committed:
Add vehicle destroy events to alert sim tool

ryanjsims committed:
added --rounds option for number of rounds to run

Commit:
Linting
